File: bnovate/bnovate/report/breakbulk_invoice_export/breakbulk_invoice_export.js
// Copyright (c) 2016, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Breakbulk Invoice Export"] = {
	filters: [
		{
			fieldname: "mawb",
			label: __("Breakbulk Master No"),
			fieldtype: "Select",
			options: [], // Initially empty, populated on load
		}, {
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1
		}
	],
	async onload(report) {
		this.report = report;

		// Populate Breakbulk Master No filter
		const mawbs = await bnovate.utils.get_mawbs();
		const options = [""].concat(mawbs);
		report.get_filter("mawb").df.options = options;
		report.get_filter("mawb").value = mawbs[0] || "";
		report.get_filter("mawb").refresh();

		report.page.add_inner_button(__('<i class="fa fa-download"></i> Export to CSV'), () => {
			const column_row = this.report.columns.map(col => col.label);
			const data = this.report.get_data_for_csv();
			const out = [column_row].concat(data);

			frappe.tools.downloadify(out, null, this.report.report_name);
		});

		report.page.add_inner_button(__('<i class="fa fa-download"></i> Commercial Invoices'), () => {
			const dn_list = Array.from(new Set(frappe.query_report.data.map(row => row.dn_name)));
			bnovate.utils.open_pdf_urls("Delivery Note", dn_list, "Commercial Invoice CHF");
		});

		report.page.add_inner_button(__('<i class="fa fa-calculator"></i> Sum Shipping'), () => {
			let shipping = frappe.query_report.data
				.filter(row => row.item_name === 'Shipping')
				.reduce((acc, row) => acc + row.base_declared_amount, 0);

			bnovate.utils.html_dialog('Total shipping charged',
				`Sum of shipping rows:<br><br> <b>${shipping.toFixed(2)}</b> CHF`);
		})
	},
};

File: bnovate/bnovate/report/breakbulk_invoice_export/breakbulk_invoice_export.py
# Copyright (c) 2013, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from erpnext import get_company_currency

from bnovate.bnovate.report.breakbulk_shipment_export.breakbulk_shipment_export import get_data as get_shipment_data


def execute(filters=None):
    columns, data = [], []
    return get_columns(), get_data(filters)


def get_columns():
    return [
        {
            "label": "Line No",
            "fieldname": "dni_idx",
            "fieldtype": "Int",
            "width": 80
        },
        {
            "label": "Agr,ment",
            "fieldname": "agreement",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": "Invoice Number",
            "fieldname": "dn_name",
            "fieldtype": "Link",
            "options": "Delivery Note",
            "width": 150
        },
        {
            "label": "Invoice Date",
            "fieldname": "posting_date",
            "fieldtype": "Data",
            "width": 120

        },
        {
            "label": "Code Client",
            "fieldname": "customer_number",
            "fieldtype": "Data",
            "width": 100
        },
        {
            "label": "Total Gross Weight",
            "fieldname": "total_gross_weight",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": "Total Net Weight",
            "fieldname": "total_net_weight",
            "fieldtype": "Float",
            "width": 120
        },
        # {
        #     "label": "EORI",
        #     "fieldname": "eori_number",
        #     "fieldtype": "Data",
        #     "width": 150
        # },
        {
            "label": "Recipient's VAT Number",
            "fieldname": "tax_id",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Recipient's Country",
            "fieldname": "billing_country",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Item Name",
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": "Customs Tariff Number",
            "fieldname": "customs_tariff_number",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Net Weight",
            "fieldname": "dni_total_weight",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": "Origin",
            "fieldname": "country_of_origin",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Preferential Origin",
            "fieldname": "preferential_origin",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": "Quantity Per Item",
            "fieldname": "qty",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": "Items Value",
            "fieldname": "base_declared_amount",
            "fieldtype": "Float",
            "width": 120
        },
        {
            "label": "Currency",
            "fieldname": "currency",
            "fieldtype": "Data",
            "width": 80
        },
        {
            "label": "AWB Number",
            "fieldname": "breakbulk_master_no",
            "fieldtype": "Data",
            "width": 150
        }
    ]

def get_data(filters):

    # Fetch data from other report to ensure consistant filtering / value calculation.
    data = get_shipment_data(filters, include_parcels=False, include_shipping=True)

    # Get gross weight per breakbulk. Count each DN only once
    unique_weights = list(set((r['breakbulk_master_no'], r['dn_name'], r['dn_total_gross_weight']) for r in data))
    
    # Add ACL-specific data
    company_currency = get_company_currency(filters.company)
    for row in data:
        row.update({
            'agreement': 'S.',
            'customer_number': '',
            'preferential_origin': 'NO',
            'total_gross_weight': sum(el[2] for el in unique_weights if el[0] == row['breakbulk_master_no']),
            'total_net_weight': sum(r['dni_total_weight'] for r in data if r['breakbulk_master_no'] == row['breakbulk_master_no']),
        })

    return data


        
